CMakeLists.txt

```diff
@@ -8,20 +8,41 @@ foreach(var ${vars})
   if("${currentHelpString}" MATCHES "No help, variable specified on the command line." OR "${currentHelpString}" STREQUAL "")
     list(APPEND CL_ARGS "-D${var}=${${var}}")
   endif()
+  if(("${var}" MATCHES "CMAKE_.*") AND NOT ("${var}" MATCHES "CMAKE_BUILD_TYPE") AND NOT ("${var}" MATCHES "CMAKE_HOME_DIRECTORY"))
+    list(APPEND PLATFORM_ARGS "-D${var}=${${var}}")
+  endif()
 endforeach()
 
 include(ExternalProject)
 project(LEAN CXX C)
 
+if("${CMAKE_SYSTEM_NAME}" MATCHES "Emscripten")
+  # For Emscripten, we build GMP before any of the stages and reuse it in all of them.
+  set(GMP_INSTALL_PREFIX ${CMAKE_BINARY_DIR}/gmp-root)
+  set(EMSCRIPTEN_SETTINGS "-s ALLOW_MEMORY_GROWTH=1")
+  ExternalProject_Add(
+    gmp
+    URL https://gmplib.org/download/gmp/gmp-6.2.1.tar.bz2
+    URL_HASH SHA256=eae9326beb4158c386e39a356818031bd28f3124cf915f8c5b1dc4c7a36b4d7c
+    BUILD_IN_SOURCE 1
+    CONFIGURE_COMMAND emconfigure ./configure "CFLAGS=${EMSCRIPTEN_SETTINGS}" --host=wasm32-unknown-emscripten --disable-assembly --prefix=${GMP_INSTALL_PREFIX}
+    BUILD_COMMAND emmake make -j4
+    INSTALL_COMMAND emmake make install
+  )
+  set(EXTRA_DEPENDS "gmp")
+  list(APPEND CL_ARGS "-DGMP_INSTALL_PREFIX=${GMP_INSTALL_PREFIX}")
+  list(APPEND PLATFORM_ARGS "-DGMP_INSTALL_PREFIX=${GMP_INSTALL_PREFIX}")
+endif()
 ExternalProject_add(stage0
   SOURCE_DIR "${LEAN_SOURCE_DIR}/stage0"
   SOURCE_SUBDIR src
   BINARY_DIR stage0
-  # always use standard release build (discarding LEAN_CXX_EXTRA_FLAGS etc.) for stage0 since it is assumed to be "good"
+  # always use standard release build (discarding LEAN_CXX_EXTRA_FLAGS etc.) for stage0 since it is assumed to be "good", but still pass through CMake platform arguments (compiler, toolchain file, ..)
```

Comment:
So stage 0 is compiled to WASM as well? How do we build the stdlib then?

Reply:
Yep, in fact for building 32-bit stdlib .oleans we *need* a 32-bit stage 0. The WASM stage 0 runs under Node.js using `node lean.js` and is basically just a normal command-line Lean (we mount the working directory as a virtual NODEFS filesystem).

Reply:
I see. So what is the status here? What works, what's still missing?

Reply:
Stage 0 compilation is working, but the stdlib target fails at the interpreter trying to call `dlsym`. I am fixing that now. When we can compile all of Lean + stdlib to a WASM that works in the browser, I will mark this as ready for review and ping you.